### PR TITLE
Refuses to execute sh or exe files

### LINKED ISSUE #99: prevent harmful code injection
As a developer, I want to produce program that is safe to have and use.
As a user, I want to be able to trust the software I'm using.

### PR DESCRIPTION
**Pull request by veikkahonkanen**
Resolves #99

### CODE REVIEW

File: src/classes/exec-sync-utils.ts [hunk 1]
```diff
@@ -1,15 +1,26 @@
...
 export const runExecSync = async (command: string, options?: ExecSyncOptionsWithBufferEncoding): Promise<string | void> => {
   try {
-    return execSync(command, options)?.toString(); //NOSONAR
+    /**
+     * Deny executable, shell, installation, registry and python files,
+     * files that are run in a sandbox without checks as well as files with possible malware included
+     */
+    if (command.search(/;|\.exe$|\.com$|\.dll$|\.vbs$|\.vb$|\.vbe$|\.pif$|\.application$|\.gadget$|\.ws$|\.wsf$|\.reg$/) == -1) {
+      if (command.search(/\.msi$|\.msp$|\.scr$|\.hta$|\.cpl$|\.msc$|\.jar$|\.bat$|\.cmd$|\.js$|\.jse$|\.wsc$|\.wsh$/) == -1) {
+        if (command.search(/\.ps.$|\.ps.xml$|\.psc.$|\.msh$|\.msh.$|\.mshxml$|\.msh.xml$|\.pdf$|\.sfx$|\.tmp$|\.py$/) == -1) {
```
Comment (eetupur) [inline src/classes/exec-sync-utils.ts]: Should probably stop execution of command this was called in. Not executing 1 or more execSyncs will cause problems if any are to be driven after.

Also these would be better as variables, we might find more to add to this list of things to block.

**Review by eetupur — CHANGES_REQUESTED** [review]
found things to change

File: src/classes/exec-sync-utils.ts (hunk 1 above)
Comment (veikkahonkanen) [reply src/classes/exec-sync-utils.ts]: File extensions have now been added into an array of regular expressions. Added errors.

File: src/classes/exec-sync-utils.ts [hunk 2]
```diff
@@ -1,15 +1,51 @@
...
+ * An array of file extensions and other parts of string that are not supposed to be found in executable strings
+ */
+const fileExtensions : RegExp[] = [
+  /;/, /\.exe$/, /\.com$/, /\.dll$/, /\.vbs$/, /\.vb$/, /\.vbe$/, /\.pif$/, /\.application$/, /\.gadget$/, /\.ws$/, /\.wsf$/,/\.reg$/,
+  /\.msi$/, /\.msp$/, /\.scr$/, /\.hta$/, /\.cpl$/, /\.msc$/, /\.jar$/, /\.bat$/, /\.cmd$/, /\.js$/, /\.jse$/, /\.wsc$/, /\.wsh$/,
+  /\.ps.$/, /\.ps.xml$/, /\.psc.$/, /\.msh$/, /\.msh.$/, /\.mshxml$/, /\.msh.xml$/, /\.pdf$/, /\.sfx$/, /\.tmp$/, /\.py$/
+];
+
+/**
+ * runs the given command in a shell, denies attempts of injection
```
Comment (jarinykanen) [inline src/classes/exec-sync-utils.ts]:
```suggestion
 * Runs the given command in a shell, denies attempts of injection
```

File: src/classes/exec-sync-utils.ts [hunk 3]
```diff
@@ -1,15 +1,51 @@
...
  * 
  * @param {string} command command to be run in shell
  * @param {ExecSyncOptionsWithBufferEncoding} options options to be provided to the shell
  * @returns {Promise<string | void>} if command outputs, will provide it as string
  */
 export const runExecSync = async (command: string, options?: ExecSyncOptionsWithBufferEncoding): Promise<string | void> => {
+  let harmful = false;
+  let foundExt : string = null;
```
Comment (jarinykanen) [inline src/classes/exec-sync-utils.ts]:
```suggestion
  let harmful: boolean = false;
  let foundExt: string = null;
```

File: src/classes/exec-sync-utils.ts [hunk 4]
```diff
@@ -1,15 +1,51 @@
...
+        harmful = true;
+        foundExt = ext;
+      }
+    }
+
+    /**
+     * Run safe commands in the shell only
+     */
+    if (!harmful) {
+      return execSync(command, options)?.toString(); //NOSONAR
```
Comment (jarinykanen) [inline src/classes/exec-sync-utils.ts:39]: NOSONAR?

**Review by jarinykanen — CHANGES_REQUESTED** [review]

File: src/classes/exec-sync-utils.ts (hunk 4 above)
Comment (veikkahonkanen) [reply src/classes/exec-sync-utils.ts:39]: Yes, this is from a previous feature

**Review by eetupur — APPROVED** [review]

File: src/classes/exec-sync-utils.ts [hunk 5]
```diff
@@ -1,15 +1,51 @@
...
+
+    } else {
+      if (foundExt == ";") {
+        throw new Error(`There was an attempt to execute extra commands.`);
+      } else {
+        throw new Error(`Found out that there was an attempt to execute ${foundExt} file.`);
```
Comment (jarinykanen) [inline src/classes/exec-sync-utils.ts]: This should return Promise.reject

**Review by jarinykanen — CHANGES_REQUESTED** [review]

File: src/classes/exec-sync-utils.ts (hunk 5 above)
Comment (veikkahonkanen) [reply src/classes/exec-sync-utils.ts]: Now returns Promise.reject()

**Review by eetupur — APPROVED** [review]

**Review by jarinykanen — APPROVED** [review]

**Pull request merged**